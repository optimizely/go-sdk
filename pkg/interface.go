/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package pkg //
package pkg

import (
	"github.com/optimizely/go-sdk/pkg/entities"
	"github.com/optimizely/go-sdk/pkg/notification"
)

// ProjectConfig contains the parsed project entities
type ProjectConfig interface {
	GetAccountID() string
	GetAnonymizeIP() bool
	GetAttributeID(id string) string // returns "" if there is no id
	GetAttributeByKey(key string) (entities.Attribute, error)
	GetAudienceByID(string) (entities.Audience, error)
	GetAudienceMap() map[string]entities.Audience
	GetBotFiltering() bool
	GetEventByKey(string) (entities.Event, error)
	GetExperimentByKey(string) (entities.Experiment, error)
	GetFeatureByKey(string) (entities.Feature, error)
	GetVariableByKey(featureKey string, variableKey string) (entities.Variable, error)
	GetFeatureList() []entities.Feature
	GetGroupByID(string) (entities.Group, error)
	GetProjectID() string
	GetRevision() string
}

// ProjectConfigManager manages the config
type ProjectConfigManager interface {
	GetConfig() (ProjectConfig, error)
	RemoveOnProjectConfigUpdate(id int) error
	OnProjectConfigUpdate(callback func(notification.ProjectConfigUpdateNotification)) (int, error)
}
