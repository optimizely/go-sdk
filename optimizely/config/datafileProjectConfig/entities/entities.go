/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

package entities

// Audience represents an Audience object from the Optimizely datafile
type Audience struct {
	ID         string      `json:"id"`
	Name       string      `json:"name"`
	Conditions interface{} `json:"condition"`
}

// Experiment represents an Experiment object from the Optimizely datafile
type Experiment struct {
	// @TODO(mng): include audienceConditions
	ID                string              `json:"id"`
	Key               string              `json:"key"`
	Status            string              `json:"status"`
	Variations        []Variation         `json:"variations"`
	TrafficAllocation []trafficAllocation `json:"trafficAllocation"`
	AudienceIds       []string            `json:"audienceIds"`
	ForcedVariations  map[string]string   `json:"forcedVariations"`
}

// FeatureFlag represents a FeatureFlag object from the Optimizely datafile
type FeatureFlag struct {
	ID            string   `json:"id"`
	RolloutID     string   `json:"rolloutId"`
	Key           string   `json:"key"`
	ExperimentIDs []string `json:"experimentIds"`
	Variables     []string `json:"variables"`
}

// trafficAllocation represents a traffic allocation range from the Optimizely datafile
type trafficAllocation struct {
	EntityID   string `json:"entityId"`
	EndOfRange int    `json:"endOfRange"`
}

// Variation represents an experiment variation from the Optimizely datafile
type Variation struct {
	ID string `json:"id"`
	// @TODO(mng): include variables
	Key            string `json:"key"`
	FeatureEnabled bool   `json:"featureEnabled"`
}

// Datafile represents the datafile we get from Optimizely
type Datafile struct {
	AccountID    string        `json:"accountId"`
	AnonymizeIP  bool          `json:"anonymizeIP"`
	Audiences    []Audience    `json:"audiences"`
	BotFiltering bool          `json:"botFiltering"`
	Experiments  []Experiment  `json:"experiments"`
	FeatureFlags []FeatureFlag `json:"featureFlags"`
	ProjectID    string        `json:"projectId"`
	Revision     string        `json:"revision"`
	Variables    []string      `json:"variables"`
	Version      string        `json:"version"`
}
