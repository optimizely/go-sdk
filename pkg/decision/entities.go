/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package decision //
package decision

import (
	"github.com/optimizely/go-sdk/pkg"
	"github.com/optimizely/go-sdk/pkg/decision/reasons"
	"github.com/optimizely/go-sdk/pkg/entities"
)

// ExperimentDecisionContext contains the information needed to be able to make a decision for a given experiment
type ExperimentDecisionContext struct {
	Experiment    *entities.Experiment
	ProjectConfig pkg.ProjectConfig
}

// FeatureDecisionContext contains the information needed to be able to make a decision for a given feature
type FeatureDecisionContext struct {
	Feature       *entities.Feature
	ProjectConfig pkg.ProjectConfig
}

// Source is where the decision came from
type Source string

const (
	// Rollout - the decision came from a rollout
	Rollout Source = "rollout"
	// FeatureTest - the decision came from a feature test
	FeatureTest Source = "feature-test"
)

// Decision contains base information about a decision
type Decision struct {
	Reason reasons.Reason
}

// FeatureDecision contains the decision information about a feature
type FeatureDecision struct {
	Decision
	Source     Source
	Experiment entities.Experiment
	Variation  *entities.Variation
}

// ExperimentDecision contains the decision information about an experiment
type ExperimentDecision struct {
	Decision
	Variation *entities.Variation
}
