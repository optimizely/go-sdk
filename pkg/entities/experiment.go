/****************************************************************************
 * Copyright 2019,2021-2025 Optimizely, Inc. and contributors                   *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package entities //
package entities

// Variation represents a variation in the experiment
type Variation struct {
	ID             string
	Variables      map[string]VariationVariable
	Key            string
	FeatureEnabled bool
}

// Cmab represents the Contextual Multi-Armed Bandit configuration for an experiment
type Cmab struct {
	AttributeIds      []string `json:"attributes"`
	TrafficAllocation int      `json:"trafficAllocation"`
}

// Experiment represents an experiment
type Experiment struct {
	AudienceIds           []string
	AudienceConditions    interface{}
	ID                    string
	LayerID               string
	Key                   string
	Variations            map[string]Variation // keyed by variation ID
	VariationKeyToIDMap   map[string]string
	TrafficAllocation     []Range
	GroupID               string
	AudienceConditionTree *TreeNode
	Whitelist             map[string]string
	IsFeatureExperiment   bool
	Cmab                  *Cmab
}

// Range represents bucketing range that the specify entityID falls into
type Range struct {
	EntityID   string
	EndOfRange int
}

// VariationVariable represents a Variable object from the Variation
type VariationVariable struct {
	ID    string
	Value string
}

// HoldoutStatus represents the status of a holdout
type HoldoutStatus string

const (
	// HoldoutStatusRunning - the holdout status is running
	HoldoutStatusRunning HoldoutStatus = "Running"
)

// Holdout represents a holdout that can be applied to feature flags
type Holdout struct {
	ID            string
	Key           string
	Status        HoldoutStatus
	IncludedFlags []string // Flag IDs this holdout specifically includes
	ExcludedFlags []string // Flag IDs this holdout specifically excludes
}
