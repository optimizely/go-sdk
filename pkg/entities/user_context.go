/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package entities //
package entities

import (
	"fmt"

	"github.com/optimizely/go-sdk/pkg/utils"
)

const bucketingIDAttributeName = "$opt_bucketing_id"

// UserContext holds information about a user
type UserContext struct {
	ID         string
	Attributes map[string]interface{}
}

// GetStringAttribute returns the string value for the specified attribute name in the attributes map. Returns error if not found.
func (u UserContext) GetStringAttribute(attrName string) (string, error) {
	if value, ok := u.Attributes[attrName]; ok {
		stringVal, err := utils.GetStringValue(value)
		if err == nil {
			return stringVal, nil
		}
	}

	return "", fmt.Errorf(`no string attribute named "%s"`, attrName)
}

// GetBoolAttribute returns the bool value for the specified attribute name in the attributes map. Returns error if not found.
func (u UserContext) GetBoolAttribute(attrName string) (bool, error) {
	if value, ok := u.Attributes[attrName]; ok {
		boolVal, err := utils.GetBoolValue(value)
		if err == nil {
			return boolVal, nil
		}
	}

	return false, fmt.Errorf(`no bool attribute named "%s"`, attrName)
}

// GetFloatAttribute returns the float64 value for the specified attribute name in the attributes map. Returns error if not found.
func (u UserContext) GetFloatAttribute(attrName string) (float64, error) {
	if value, ok := u.Attributes[attrName]; ok {
		floatVal, err := utils.GetFloatValue(value)
		if err == nil {
			return floatVal, nil
		}
	}

	return 0, fmt.Errorf(`no float attribute named "%s"`, attrName)
}

// GetIntAttribute returns the int64 value for the specified attribute name in the attributes map. Returns error if not found.
func (u UserContext) GetIntAttribute(attrName string) (int64, error) {
	if value, ok := u.Attributes[attrName]; ok {
		intVal, err := utils.GetIntValue(value)
		if err == nil {
			return intVal, nil
		}
	}

	return 0, fmt.Errorf(`no int attribute named "%s"`, attrName)
}

// GetBucketingID returns the bucketing ID to use for the given user
func (u UserContext) GetBucketingID() (string, error) {
	// by default
	bucketingID := u.ID

	// If the bucketing ID key is defined in attributes, than use that in place of the user ID
	if value, ok := u.Attributes[bucketingIDAttributeName]; ok {
		customBucketingID, err := u.GetStringAttribute(bucketingIDAttributeName)
		if err != nil {
			return bucketingID, fmt.Errorf(`invalid bucketing ID provided: "%v"`, value)
		}
		bucketingID = customBucketingID
	}

	return bucketingID, nil
}
