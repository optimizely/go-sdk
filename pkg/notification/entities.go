/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package notification //
package notification

import (
	"github.com/optimizely/go-sdk/pkg/entities"
)

// Type is the type of notification
type Type string

// DecisionNotificationType is the type of decision notification
type DecisionNotificationType string

const (
	// Decision notification type
	Decision Type = "decision"
	// ProjectConfigUpdate notification type
	ProjectConfigUpdate Type = "project_config_update"

	// ABTest is used when the decision is returned as part of evaluating an ab test
	ABTest DecisionNotificationType = "ab-test"
	// Feature is used when the decision is returned as part of evaluating a feature
	Feature DecisionNotificationType = "feature"
	// LogEvent notification type
	LogEvent Type = "log_event_notification"
)

// DecisionNotification is a notification triggered when a decision is made for either a feature or an experiment
type DecisionNotification struct {
	Type         DecisionNotificationType
	UserContext  entities.UserContext
	DecisionInfo map[string]interface{}
}

// ProjectConfigUpdateNotification is a notification triggered when a project config is updated
type ProjectConfigUpdateNotification struct {
	Type     Type
	Revision string
}

// LogEventNotification is the notification triggered before log event is dispatched.
type LogEventNotification struct {
	Type         Type
	LogEvent  	 interface{}
}
