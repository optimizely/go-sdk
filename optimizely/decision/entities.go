/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

package decision

import "github.com/optimizely/go-sdk/optimizely/entities"

// ExperimentDecisionContext contains the information needed to be able to make a decision for a given experiment
type ExperimentDecisionContext struct {
	AudienceMap map[string]entities.Audience
	Experiment  entities.Experiment
	Group       entities.Group
}

// FeatureDecisionContext contains the information needed to be able to make a decision for a given feature
type FeatureDecisionContext struct {
	Feature entities.Feature
	Group   entities.Group
}

// Decision contains base information about a decision
type Decision struct {
	DecisionMade bool
}

// FeatureDecision contains the decision information about a feature
type FeatureDecision struct {
	Decision
	FeatureEnabled bool
}

// ExperimentDecision contains the decision information about an experiment
type ExperimentDecision struct {
	Decision
	Variation entities.Variation
}
