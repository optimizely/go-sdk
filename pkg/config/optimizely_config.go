/****************************************************************************
 * Copyright 2019, Optimizely, Inc. and contributors                        *
 *                                                                          *
 * Licensed under the Apache License, Version 2.0 (the "License");          *
 * you may not use this file except in compliance with the License.         *
 * You may obtain a copy of the License at                                  *
 *                                                                          *
 *    http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                          *
 * Unless required by applicable law or agreed to in writing, software      *
 * distributed under the License is distributed on an "AS IS" BASIS,        *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
 * See the License for the specific language governing permissions and      *
 * limitations under the License.                                           *
 ***************************************************************************/

// Package config //
package config

import "github.com/optimizely/go-sdk/pkg/entities"

// OptimizelyConfig is a snapshot of the experiments and features in the project config
type OptimizelyConfig struct {
	Revision       string
	ExperimentsMap map[string]OptimizelyExperiment
	FeaturesMap    map[string]OptimizelyFeature
}

// OptimizelyExperiment has experiment info
type OptimizelyExperiment struct {
	ID            string
	Key           string
	VariationsMap map[string]OptimizelyVariation
}

// OptimizelyFeature has feature info
type OptimizelyFeature struct {
	ID             string
	Key            string
	ExperimentsMap map[string]OptimizelyExperiment
	VariablesMap   map[string]OptimizelyVariable
}

// OptimizelyVariation has variation info
type OptimizelyVariation struct {
	ID             string
	Key            string
	FeatureEnabled bool
	VariablesMap   map[string]OptimizelyVariable
}

// OptimizelyVariable has variable info
type OptimizelyVariable struct {
	ID    string
	Key   string
	Type  string
	Value string
}

func getVariableByIDMap(features []entities.Feature) (variableByIDMap map[string]entities.Variable) {
	variableByIDMap = map[string]entities.Variable{}
	for _, feature := range features {
		for _, variable := range feature.VariableMap {
			variableByIDMap[variable.ID] = variable
		}
	}
	return variableByIDMap
}

func getExperimentVariablesMap(features []entities.Feature) (experimentVariableMap map[string]map[string]OptimizelyVariable) {
	experimentVariableMap = map[string]map[string]OptimizelyVariable{}
	for _, feature := range features {

		var optimizelyVariableMap = map[string]OptimizelyVariable{}
		for _, variable := range feature.VariableMap {
			optimizelyVariableMap[variable.Key] = OptimizelyVariable{Key: variable.Key, ID: variable.ID, Value: variable.DefaultValue, Type: string(variable.Type)}

		}
		for _, experiment := range feature.FeatureExperiments {
			experimentVariableMap[experiment.Key] = optimizelyVariableMap
		}
	}
	return experimentVariableMap
}

func getExperimentMap(features []entities.Feature, experiments []entities.Experiment, variableByIDMap map[string]entities.Variable) (optlyExperimentMap map[string]OptimizelyExperiment) {

	optlyExperimentMap = map[string]OptimizelyExperiment{}
	experimentVariablesMap := getExperimentVariablesMap(features)

	for _, experiment := range experiments {
		var optlyVariationsMap = map[string]OptimizelyVariation{}
		for _, variation := range experiment.Variations {
			var optlyVariablesMap = map[string]OptimizelyVariable{}

			if variableMap, ok := experimentVariablesMap[experiment.Key]; ok {
				for index, element := range variableMap { // copy by value
					optlyVariablesMap[index] = element
				}
			}

			for _, variable := range variation.Variables {
				if experiment.IsFeatureExperiment && variation.FeatureEnabled {
					if convertedVariable, ok := variableByIDMap[variable.ID]; ok {
						optlyVariable := OptimizelyVariable{Key: convertedVariable.Key, ID: convertedVariable.ID,
							Type: string(convertedVariable.Type), Value: variable.Value}
						optlyVariablesMap[convertedVariable.Key] = optlyVariable
					}
				}
			}
			optVariation := OptimizelyVariation{ID: variation.ID, Key: variation.Key, VariablesMap: optlyVariablesMap, FeatureEnabled: variation.FeatureEnabled}
			optlyVariationsMap[variation.Key] = optVariation
		}
		optlyExperiment := OptimizelyExperiment{ID: experiment.ID, Key: experiment.Key, VariationsMap: optlyVariationsMap}
		optlyExperimentMap[experiment.Key] = optlyExperiment
	}
	return optlyExperimentMap
}

func getFeatureMap(features []entities.Feature, experimentsMap map[string]OptimizelyExperiment) (optlyFeatureMap map[string]OptimizelyFeature) {

	optlyFeatureMap = map[string]OptimizelyFeature{}

	for _, feature := range features {

		var optlyFeatureVariablesMap = map[string]OptimizelyVariable{}
		for _, featureVarible := range feature.VariableMap {
			optlyVariable := OptimizelyVariable{Key: featureVarible.Key, ID: featureVarible.ID,
				Type: string(featureVarible.Type), Value: featureVarible.DefaultValue}
			optlyFeatureVariablesMap[featureVarible.Key] = optlyVariable
		}

		var optlyExperimentMap = map[string]OptimizelyExperiment{}
		for _, experiment := range feature.FeatureExperiments {
			optlyExperimentMap[experiment.Key] = experimentsMap[experiment.Key]
		}

		optlyFeature := OptimizelyFeature{ID: feature.ID, Key: feature.Key, ExperimentsMap: optlyExperimentMap, VariablesMap: optlyFeatureVariablesMap}
		optlyFeatureMap[feature.Key] = optlyFeature

	}
	return optlyFeatureMap
}

// NewOptimizelyConfig constructs OptimizelyConfig object
func NewOptimizelyConfig(projConfig ProjectConfig) *OptimizelyConfig {

	if projConfig == nil {
		return nil
	}
	featuresList := projConfig.GetFeatureList()
	experimentsList := projConfig.GetExperimentList()
	revision := projConfig.GetRevision()

	optimizelyConfig := &OptimizelyConfig{}

	variableByIDMap := getVariableByIDMap(featuresList)

	optimizelyConfig.ExperimentsMap = getExperimentMap(featuresList, experimentsList, variableByIDMap)
	optimizelyConfig.FeaturesMap = getFeatureMap(featuresList, optimizelyConfig.ExperimentsMap)
	optimizelyConfig.Revision = revision

	return optimizelyConfig
}
