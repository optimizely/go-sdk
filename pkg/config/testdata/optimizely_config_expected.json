{
  "environmentKey": "",
  "sdkKey": "",
  "revision": "242",
  "experimentsMap": {
    "ab_running_exp_single_exact_match_string_typedaudience": {
      "id": "10390977674",
      "key": "ab_running_exp_single_exact_match_string_typedaudience",
      "audiences": "",
      "variationsMap": {
        "all_traffic_variation": {
          "id": "10416523121",
          "key": "all_traffic_variation",
          "featureEnabled": false,
          "variablesMap": {
            "b_true": {
              "id": "26874700961",
              "key": "b_true",
              "type": "boolean",
              "value": "true"
            },
            "d_4_2": {
              "id": "26874700951",
              "key": "d_4_2",
              "type": "double",
              "value": "42.2"
            },
            "i_42": {
              "id": "26874700941",
              "key": "i_42",
              "type": "integer",
              "value": "42"
            },
            "s_foo": {
              "id": "26874700971",
              "key": "s_foo",
              "type": "string",
              "value": "foo"
            }
          }
        }
      }
    },
    "ab_running_exp_single_exact_match_string_untypedaudience": {
      "id": "10390977673",
      "key": "ab_running_exp_single_exact_match_string_untypedaudience",
      "audiences": "",
      "variationsMap": {
        "all_traffic_variation": {
          "id": "10416523121",
          "key": "all_traffic_variation",
          "featureEnabled": false,
          "variablesMap": {
            "b_true": {
              "id": "26874700961",
              "key": "b_true",
              "type": "boolean",
              "value": "true"
            },
            "d_4_2": {
              "id": "26874700951",
              "key": "d_4_2",
              "type": "double",
              "value": "42.2"
            },
            "i_42": {
              "id": "26874700941",
              "key": "i_42",
              "type": "integer",
              "value": "42"
            },
            "s_foo": {
              "id": "26874700971",
              "key": "s_foo",
              "type": "string",
              "value": "foo"
            }
          }
        }
      }
    }
  },
  "featuresMap": {
    "ab_running_exp_single_exact_match_string_typedaudience": {
      "id": "1000000017",
      "key": "ab_running_exp_single_exact_match_string_typedaudience",
      "experimentRules": [
        {
          "id": "10390977674",
          "key": "ab_running_exp_single_exact_match_string_typedaudience",
          "audiences": "",
          "variationsMap": {
            "all_traffic_variation": {
              "id": "10416523121",
              "key": "all_traffic_variation",
              "featureEnabled": false,
              "variablesMap": {
                "b_true": {
                  "id": "26874700961",
                  "key": "b_true",
                  "type": "boolean",
                  "value": "true"
                },
                "d_4_2": {
                  "id": "26874700951",
                  "key": "d_4_2",
                  "type": "double",
                  "value": "42.2"
                },
                "i_42": {
                  "id": "26874700941",
                  "key": "i_42",
                  "type": "integer",
                  "value": "42"
                },
                "s_foo": {
                  "id": "26874700971",
                  "key": "s_foo",
                  "type": "string",
                  "value": "foo"
                }
              }
            }
          }
        }
      ],
      "deliveryRules": [],
      "variablesMap": {
        "b_true": {
          "id": "26874700961",
          "key": "b_true",
          "type": "boolean",
          "value": "true"
        },
        "d_4_2": {
          "id": "26874700951",
          "key": "d_4_2",
          "type": "double",
          "value": "42.2"
        },
        "i_42": {
          "id": "26874700941",
          "key": "i_42",
          "type": "integer",
          "value": "42"
        },
        "s_foo": {
          "id": "26874700971",
          "key": "s_foo",
          "type": "string",
          "value": "foo"
        }
      },
      "experimentsMap": {
        "ab_running_exp_single_exact_match_string_typedaudience": {
          "id": "10390977674",
          "key": "ab_running_exp_single_exact_match_string_typedaudience",
          "audiences": "",
          "variationsMap": {
            "all_traffic_variation": {
              "id": "10416523121",
              "key": "all_traffic_variation",
              "featureEnabled": false,
              "variablesMap": {
                "b_true": {
                  "id": "26874700961",
                  "key": "b_true",
                  "type": "boolean",
                  "value": "true"
                },
                "d_4_2": {
                  "id": "26874700951",
                  "key": "d_4_2",
                  "type": "double",
                  "value": "42.2"
                },
                "i_42": {
                  "id": "26874700941",
                  "key": "i_42",
                  "type": "integer",
                  "value": "42"
                },
                "s_foo": {
                  "id": "26874700971",
                  "key": "s_foo",
                  "type": "string",
                  "value": "foo"
                }
              }
            }
          }
        }
      }
    },
    "ab_running_exp_single_exact_match_string_untypedaudience": {
      "id": "1000000016",
      "key": "ab_running_exp_single_exact_match_string_untypedaudience",
      "experimentRules": [
        {
          "id": "10390977673",
          "key": "ab_running_exp_single_exact_match_string_untypedaudience",
          "audiences": "",
          "variationsMap": {
            "all_traffic_variation": {
              "id": "10416523121",
              "key": "all_traffic_variation",
              "featureEnabled": false,
              "variablesMap": {
                "b_true": {
                  "id": "26874700961",
                  "key": "b_true",
                  "type": "boolean",
                  "value": "true"
                },
                "d_4_2": {
                  "id": "26874700951",
                  "key": "d_4_2",
                  "type": "double",
                  "value": "42.2"
                },
                "i_42": {
                  "id": "26874700941",
                  "key": "i_42",
                  "type": "integer",
                  "value": "42"
                },
                "s_foo": {
                  "id": "26874700971",
                  "key": "s_foo",
                  "type": "string",
                  "value": "foo"
                }
              }
            }
          }
        }
      ],
      "deliveryRules": [],
      "variablesMap": {
        "b_true": {
          "id": "26874700961",
          "key": "b_true",
          "type": "boolean",
          "value": "true"
        },
        "d_4_2": {
          "id": "26874700951",
          "key": "d_4_2",
          "type": "double",
          "value": "42.2"
        },
        "i_42": {
          "id": "26874700941",
          "key": "i_42",
          "type": "integer",
          "value": "42"
        },
        "s_foo": {
          "id": "26874700971",
          "key": "s_foo",
          "type": "string",
          "value": "foo"
        }
      },
      "experimentsMap": {
        "ab_running_exp_single_exact_match_string_untypedaudience": {
          "id": "10390977673",
          "key": "ab_running_exp_single_exact_match_string_untypedaudience",
          "audiences": "",
          "variationsMap": {
            "all_traffic_variation": {
              "id": "10416523121",
              "key": "all_traffic_variation",
              "featureEnabled": false,
              "variablesMap": {
                "b_true": {
                  "id": "26874700961",
                  "key": "b_true",
                  "type": "boolean",
                  "value": "true"
                },
                "d_4_2": {
                  "id": "26874700951",
                  "key": "d_4_2",
                  "type": "double",
                  "value": "42.2"
                },
                "i_42": {
                  "id": "26874700941",
                  "key": "i_42",
                  "type": "integer",
                  "value": "42"
                },
                "s_foo": {
                  "id": "26874700971",
                  "key": "s_foo",
                  "type": "string",
                  "value": "foo"
                }
              }
            }
          }
        }
      }
    }
  },
  "attributes": [
    { "id": "10401066170", "key": "string_attribute" },
    { "id": "10401066171", "key": "s_foo" }
  ],
  "audiences": [
    {
      "id": "20413101794",
      "name": "single_condition_leaf_root",
      "conditions": "{\"match\":\"exact\",\"name\":\"s_foo\",\"type\":\"custom_attribute\",\"value\":\"leaf_root\"}"
    },
    {
      "id": "10413101794",
      "name": "untyped_single_condition_exact_string_match",
      "conditions": "[\"and\", [\"or\", [\"or\", {\"type\": \"custom_attribute\", \"name\": \"string_attribute\", \"value\": \"exact_match\"}]]]"
    }
  ],
  "events": [
    {
      "id": "10404198135",
      "key": "event_single_targeted_exp",
      "experimentIds": ["10390977673"]
    },
    {
      "id": "10404198136",
      "key": "event_single_untargeted_exp",
      "experimentIds": ["10390977674"]
    }
  ]
}
